reconcile/openshift_resources.py

```diff
@@ -119,107 +175,134 @@ def fetch_provider_resource(path):
+            openshift_resource = OR(item)
+            ri.add_current(
+                cluster,
+                namespace,
+                resource_type,
+                openshift_resource.name,
+                openshift_resource
+            )
+    return ri
+
+
+def fetch_desired_state(ri, cluster, namespace, openshift_resources):
+    for resource in openshift_resources:
+        openshift_resource = fetch_openshift_resource(resource)
+        if openshift_resource is None:
+            error('Invalid None type for openshift_resource.')
+            continue
+
+        # add to inventory
+        try:
+            ri.add_desired(
+                cluster,
+                namespace,
+                openshift_resource.kind,
+                openshift_resource.name,
+                openshift_resource
+            )
+        except KeyError:
+            # This is failing because in the managed_type loop (where the
+            # `initialize_resource_type` method was called), this specific
+            # combination was not initialized, meaning that it shouldn't be
+            # managed. But someone is trying to add it via app-interface
+            error("Unknown kind: {}.".format(openshift_resource.kind))
+            continue
+        except ResourceKeyExistsError:
+            # This is failing because an attempt to add
+            # a desired resource with the same name and
+            # the same type was already added previously
+            error("Desired item already exists: {}/{}.".format(
+                openshift_resource.kind, openshift_resource.name))
+            continue
+    return ri
+
+
 def fetch_data(namespaces_query):
     ri = ResourceInventory()
     oc_map = {}
-    errors = []
 
     for namespace_info in namespaces_query:
-        namespace = namespace_info['name']
-        cluster_info = namespace_info['cluster']
-        cluster = cluster_info['name']
-
         # Skip if namespace has no managedResourceTypes
         managed_types = namespace_info.get('managedResourceTypes')
         if not managed_types:
             continue
 
-        # useful errors
-        def error(msg):
-            err_msg = (
-                "Error fetching data. cluster: {}, namespace: {}. "
-            ).format(cluster, namespace) + msg
-
-            logging.error(err_msg)
-            errors.append(err_msg)
-
-        # Obtain `oc` client
-        if oc_map.get(cluster) is None:
-            at = cluster_info.get('automationToken')
-
-            # Skip if cluster has no automationToken
-            if at is None:
-                error("Cluster {} has no automationToken.")
-                oc_map[cluster] = False
-            else:
-                token = vault_client.read(at['path'], at['field'])
-                oc_map[cluster] = OC(cluster_info['serverUrl'], token)
-
-        oc = oc_map[cluster]
-
+        cluster_info = namespace_info['cluster']
+        oc = obtain_oc_client(oc_map, cluster_info)
```

Comment:
very nice, I like this idea

Reply:
thanks!
i usually find that comments are the best indication that this should be a function